tools/istio-iptables/pkg/validation/validator.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validation
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"time"
+
+	"istio.io/istio/tools/istio-iptables/pkg/config"
+	"istio.io/istio/tools/istio-iptables/pkg/constants"
+)
+
+type Validator struct {
+	Config *Config
+}
+
+type Config struct {
+	ServerListenAddress string
+	ServerOriginalPort  uint16
+	ServerOriginalIP    net.IP
+}
+type Service struct {
+	Config *Config
+}
+type Client struct {
+	Config *Config
+}
+
+func (validator *Validator) Run() error {
+	s := Service{
+		validator.Config,
+	}
+	sError := make(chan error, 1)
+	sTimeout := time.After(30 * time.Second)
+	go func() {
+		sError <- s.Run()
+	}()
+
+	// infinite loop
+	go func() {
+		c := Client{Config: validator.Config}
+		for {
+			_ = c.Run()
+			// Avoid spamming the request to the validation server.
+			// Since the TIMEWAIT socket is cleaned up in 60 second,
+			// it's maintaining 60 TIMEWAIT sockets. Not big deal.
+			time.Sleep(time.Second)
+		}
+	}()
+	select {
+	case <-sTimeout:
+		return errors.New("validation timeout")
+	case err := <-sError:
+		if err == nil {
+			fmt.Println("validation passed!")
+		} else {
+			fmt.Println("validation failed:" + err.Error())
+		}
+		return err
+	}
+}
+
+func NewValidator(config *config.Config) *Validator {
+	return &Validator{
+		Config: &Config{
+			ServerListenAddress: ":" + config.InboundCapturePort,
+			ServerOriginalPort:  constants.IptablesProbePort,
+			ServerOriginalIP:    config.HostIP,
+		},
+	}
+}
+
+// Write human readable response
+func echo(conn io.WriteCloser, echo []byte) {
+	_, _ = conn.Write(echo)
+	_ = conn.Close()
+}
+
+func (s *Service) Run() error {
+	l, err := net.Listen("tcp", s.Config.ServerListenAddress)
```

Comment:
Are you testing both IPv4 and IPv6 when IPv6 is disabled? If not, why?

Reply:
The server address `s.Config.ServerListenAddress` is `::15006`, which includes both tcpv6 and tcpv4.
As a service it should listen on both v6 and v4 as long as it can.

The testing strategy decide if we test only v4, or only v6, or both.

In this PR, I believe it's testing either v4 or v6. The goal is to assume iptables setup is atomic: either none or all. 

Should I assume v4 and v6 are failing independently?

Reply:
Now that we've moved to using `iptables-restore` and `ip6tables-restore`, I would assume that IPv4 is setup atomically and IPv6 is setup atomically. But both together are not set atomically.
Your call. At least, this implementation does the job of testing whether Istio CNI setup IPv4 redirection successfully. That's the minimum.

Reply:
Got it. The goal for this PR is to detect ipv4 failure.  
To be clear: if the user configures ipv4 iptables the failure detector will detect it. Otherwise failure detector is noop.

Will look into we are intercepting v4, v6 or both

Reply:
Update: Currently istio iptables chooses one from ipv4 and ipv6, and forbid, which is not ideal.
That is not ideal. But this PRs will follow this rule.